core/src/main/java/org/jboss/jandex/EquivalenceKey.java

```diff
@@ -0,0 +1,571 @@
+            case WILDCARD_TYPE:
+                return new WildcardTypeEquivalenceKey(of(type.asWildcardType().bound()), type.asWildcardType().isExtends(),
+                        type.asWildcardType().hasImplicitObjectBound());
+            default:
+                throw new AssertionError("Unknown type " + type);
+        }
+    }
+
+    private static TypeEquivalenceKey[] of(Type[] types) {
+        TypeEquivalenceKey[] result = new TypeEquivalenceKey[types.length];
+        for (int i = 0; i < types.length; i++) {
+            result[i] = of(types[i]);
+        }
+        return result;
+    }
+
+    // ---
+
+    private EquivalenceKey() {
+    }
+
+    public static abstract class DeclarationEquivalenceKey extends EquivalenceKey {
+    }
+
+    public static final class ClassEquivalenceKey extends DeclarationEquivalenceKey {
+        private final DotName className;
+
+        private ClassEquivalenceKey(DotName className) {
+            this.className = className;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (!(o instanceof ClassEquivalenceKey))
+                return false;
+            ClassEquivalenceKey key = (ClassEquivalenceKey) o;
+            return Objects.equals(className, key.className);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(className);
+        }
+
+        @Override
+        public String toString() {
+            return "class " + className.toString();
+        }
+    }
+
+    public static final class MethodEquivalenceKey extends DeclarationEquivalenceKey {
+        private final DotName className;
+        private final byte[] methodName;
+        private final TypeEquivalenceKey[] parameterTypes;
+        private final TypeEquivalenceKey returnType; // needed to distinguish bridge methods from "real" methods
+
+        private MethodEquivalenceKey(DotName className, byte[] methodName, TypeEquivalenceKey[] parameterTypes,
+                TypeEquivalenceKey returnType) {
+            this.className = className;
+            this.methodName = methodName;
+            this.parameterTypes = parameterTypes;
+            this.returnType = returnType;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (!(o instanceof MethodEquivalenceKey))
+                return false;
+            MethodEquivalenceKey that = (MethodEquivalenceKey) o;
+            return Objects.equals(className, that.className)
+                    && Arrays.equals(methodName, that.methodName)
+                    && Arrays.equals(parameterTypes, that.parameterTypes)
+                    && Objects.equals(returnType, that.returnType);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = Objects.hash(className, returnType);
```

Comment:
Perf experts always say that `Objects.hash()` is evil when used in hot paths because it allocates one more unnecessary array ;-). In fact, it also sometimes helps to compute the hashcode in the constructor and save it in a field. 

It depends on whether we expect the equivalence keys in hash maps and in hot paths.

Reply:
I'm ignoring this for now. If that ever becomes problematic, it's straightforward to improve.